Upgrade to Scala.js 1.0.0-M5

sjrd opened the pull request:
~~This is completely broken at the moment, but I'm leaving tomorrow for a week, so I'll just let it sit there.~~

sjrd commented:
@gzm0 I think I'm good with `PhantomJSEnv` per se. However I'm stuck on `RetryingComJSEnv`. I *think* I've implemented it correctly, but it has become impossible to *test*: the same instance of `JSEnv` is reused across all tests in the test suite, so we cannot keep track of when we should fail and when we should not in isolation. This is particularly problematic for the test [`multiEnvTest`](https://github.com/scala-js/scala-js/blob/9e4ae15954a16e70ee9c859838e1db9557b78780/js-envs-test-kit/src/main/scala/org/scalajs/jsenv/test/ComTests.scala#L61-L80).

I'm not even sure that it is otherwise correctly implemented.

Any idea?

sjrd commented:
Bigger problem with `RetryingComJSEnv`: it is not able to deal with non-inherited I/O. Indeed, it would call `onOutputStream` several times if it retries, which is not permitted.

I think it's not possible to implement `RetryingComJSEnv` with the new API. Should we just remove it? (Or implement it as non-retrying wrapper and deprecate it.)

gzm0 commented:
Re: RetryingEnv:

> but it has become impossible to test:

It probably is necessary to write specific tests for that functionality (rather then re-using the conformance suite). IIUC a correctly implemented (non mocked) `RetryingComJSEnv` will pass conformance without issue.

> it is not able to deal with non-inherited I/O

It seems that `RetryingComJSEnv` doesn't have the same behavior for com than for I/O (it only retries if com is reproducible, otherwise fails). If we want to retain this behavior, you'll have to use a pipe when I/O is non-inherited so you can invoke `onOutputStream` only once and pipe I/O forward from the underlying env (that will call your own `onOutputStream` again if it gets restarted).

However, given the inconsistency of definition, I'm OK with removing `RetryingComJSEnv` entirely and expect users to retry on a higher level if necessary. (Maybe we should run this against the community though :S).

sjrd commented:
> It probably is necessary to write specific tests for that functionality (rather then re-using the conformance suite). IIUC a correctly implemented (non mocked) RetryingComJSEnv will pass conformance without issue.

Ah, wow, I should have thought about that! Following that strategy, I actually got the Suite to pass on the one hand, and to write a meaningful test for the retrying behavior on the other hand.

I think this PR is in a sufficiently good shape for a review, now.

~~Question: maybe we should declare that `RetryingComJSEnv` does not support non-inherited I/O, using the Validator?~~ If we do that the Suite does not pass anymore. Some of it (the compare-output tests) requires support for non-inherited I/O ...

gzm0 commented on build.sbt:
```diff
@@ -93,7 +81,11 @@ lazy val `scalajs-env-phantomjs`: Project = project.in(file("phantomjs-env")).
+
+    parallelExecution in Test := false,
+    logBuffered in Test := false,
+    testOptions += Tests.Argument(TestFrameworks.JUnit, "-v", "-a")
```
Are these spurious?

sjrd replied:
They were for debugging, indeed. I've removed them.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -8,107 +8,80 @@
-class PhantomJSEnv(config: PhantomJSEnv.Config)
-    extends ExternalJSEnv with ComJSEnv {
-
+class PhantomJSEnv(config: PhantomJSEnv.Config) extends JSEnv {
```
final? (if we need to expose things for PhantomJS2, we should try and do so that downstream uses composition, not inheritance).

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -245,263 +231,251 @@ class PhantomJSEnv(config: PhantomJSEnv.Config)
+             |    phantom.exit(3);
+             |  } else if (data.action === 'exit') {
+             |    phantom.exit(data.returnValue || 0);
+             |  } else if (data.action === 'setAutoExit') {
```
auto exit is not needed anymore and we should remove it.

gzm0 commented on build.sbt:
```diff
@@ -1,23 +1,11 @@
-  crossScalaVersions := Seq("2.10.7", "2.11.12", "2.12.4"),
-  crossSbtVersions := Seq("1.0.4", "0.13.16"),
+  crossScalaVersions := Seq("2.10.7", "2.11.12", "2.12.6"),
+  crossSbtVersions := Seq("1.0.4", "0.13.17"),
```
Maybe split the sbt upgrade into a separate commit?

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -245,263 +231,251 @@ class PhantomJSEnv(config: PhantomJSEnv.Config)
-    override protected def initFiles(): Seq[VirtualJSFile] =
-      super.initFiles :+ comSetup
+    runConfig.logger.debug(
+        "PhantomJS using launcher at: " + launcherTmpF.getAbsolutePath())
```
Logging should be moved one level up.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -245,263 +231,251 @@ class PhantomJSEnv(config: PhantomJSEnv.Config)
+      script match {
+        /* The most horrible way to allow RunTests.syntaxErrorTest in the
+         * conformance test suite to succeed, despite #10.
+         */
```
Have you tried registering a `window.onError` handler (like Selenium does)? We really must find a way to fix this. Succeeding when unparseable code is passed is unacceptable.

sjrd replied:
I've tried *everything*. I didn't find any solution.

My least bad idea at this point is the following: for each script, load the file, wrap its contents in something like the following:
```html
<script type="text/javascript">
(function() {
  <insert script content here>
}); // do not execute the function
var SCRIPT_1_PARSES = true;
</script>
<script type="text/javascript">
if (typeof SCRIPT_1_PARSES === 'undefined')
  throw new SyntaxError("Script 1 did not parse");
</script>
```

gzm0 replied:
I think that is better than not failing on syntax tests (and excluding tests magically).

sjrd replied:
Done.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -119,37 +92,50 @@ class PhantomJSEnv(config: PhantomJSEnv.Config)
+      underlyingRun = internalStart(comSetup :: files, runConfig)
+      underlyingRun.future.onComplete { result =>
+        promise.tryComplete(result)
+      }
```
This probably needs to be synced with `mgr` (so that if the manager is not closed yet, it doesn't complete).

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -119,37 +92,50 @@ class PhantomJSEnv(config: PhantomJSEnv.Config)
-      val mgr = ctors.head.newInstance(websocketListener)
+      val listener = new WebsocketListener {
+        def onRunning(): Unit = startPhantomJSProcess()
+        def onOpen(): Unit = sendPendingMessages()
```
Please abstract this over Futures so there is less callback. It might be a good idea to adapt our shim for classloading to give access to futures (and only take an `onMessage` and `log` handle).

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -8,107 +8,80 @@
-    }
+    private val sendBuffer = mutable.ListBuffer.empty[String]
+    private var underlyingRun: JSRun = null
+    private var canSendNow: Boolean = false
```
These two scream factory or Future.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -8,107 +8,80 @@
-        mgrIsRunning = true
-        ComPhantomRunner.this.notifyAll()
-      }
+  private class ComPhantomRun(files: List[VirtualBinaryFile],
```
Might be good time to put this in a separate file? Also `final` :)

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/RetryingComJSEnv.scala:
```diff
@@ -30,165 +30,103 @@ import scala.util.{Try, Failure, Success}
-          case Success(v) => v
+      val run = baseEnv.startWithCom(input, runConfig, { msg =>
+        val doForward = synchronized {
+          if (retryCount == myRetryCount) {
```
This should not be needed. If the `future` of a run completes, the callback handler must not be called anymore. If it does, it is a bug.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/RetryingComJSEnv.scala:
```diff
@@ -30,165 +30,103 @@ import scala.util.{Try, Failure, Success}
+
+      run.future.onComplete { result =>
+        synchronized {
+          if (retryCount == myRetryCount) {
```
Same here, why do we need this?

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/RetryingComJSEnv.scala:
```diff
@@ -30,165 +30,103 @@ import scala.util.{Try, Failure, Success}
-    def stop(): Unit = ()
+  def startWithCom(input: Input, runConfig: RunConfig,
+      onMessage: String => Unit): JSComRun = {
+    new RetryingJSComRun(input, runConfig, onMessage)
```
This doesn't handle non-inherited I/O properly. TBH I think we should just drop this entire env. It is way too over engineered for what it is useful for.

sjrd replied:
#19

gzm0 commented on phantomjs-env/src/test/scala/org/scalajs/jsenv/phantomjs/RetryingComJSSuite.scala:
```diff
@@ -0,0 +1,25 @@
+  val Config = {
+    val jsEnv = new RetryingComJSEnv(new NodeJSEnv, maxFails)
+    JSEnvSuiteConfig(jsEnv)
+      .withTerminateVMJSCode("__ScalaJSEnv.exitFunction(0)")
```
It seems like we should move this to the base test suite instead (under something like `withSupportsExit`).

sjrd commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,269 @@
+  def future: Future[Unit] = promise.future
+
+  private val sendBuffer = mutable.ListBuffer.empty[String]
+  private var canSendNow: Boolean = false
```
@gzm0 said:

> These two scream factory or Future.

(only applies to `canSendNow` at this point)

sjrd replied:
I don't see how a `Future` can help, here. We can't just queue up messages to be sent as `onSuccess` handlers to the `Future`, because `Future` does not give any guarantee about the order in which the handlers are called, whereas our `send()` does.

sjrd commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,269 @@
+
+    val listener = new WebsocketListener {
+      def onRunning(): Unit = startPhantomJSProcess()
+      def onOpen(): Unit = sendPendingMessages()
```
@gzm0 said:

> Please abstract this over Futures so there is less callback. It might be a good idea to adapt our shim for classloading to give access to futures (and only take an `onMessage` and `log` handle).

sjrd replied:
I don't see how using `Future`s will result in less callbacks. If `onRunning()` performs a `runningPromise.success(())` instead of `startPhantomJSProcess()`, then I need somewhere else
```scala
runningPromise.future.onSuccess(_ => startPhantomJSProcess())
```
It does not help.

Having `Future`s cross the classloading boundary will be *insanely* hard. Every non-JDK class that needs to be loaded from the parent classloader needs to be explicitly listed in `PhantomJettyClassLoader.bridgeClasses`. Imagine the horror if we need to put `Future` and its dependencies in that list.

TBH I'm pretty sure those simple callbacks are the appropriate abstraction here, not futures.

sjrd commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,269 @@
+    val underlyingRun = startRun(comSetup)
+    underlyingRun.future.onComplete { result =>
+      promise.tryComplete(result)
+    }
```
@gzm0 said:

> This probably needs to be synced with `mgr` (so that if the manager is not closed yet, it doesn't complete).

sjrd replied:
What's the problem if the future does complete before the manager is closed?

sjrd commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -8,500 +8,266 @@
-
-      val deadline = OptDeadline(timeout)
+    runConfig.logger.debug(
+        "PhantomJS using launcher at: " + launcherTmpF.getAbsolutePath())
```
@gzm0 said:

> Logging should be moved one level up.

sjrd replied:
I don't understand what you mean, here :s

sjrd commented:
Rebased.

I currently get intermittent failures like this on Travis :s
```
[error] Test org.scalajs.jsenv.test.RunTests.allowScriptTags[config = PhantomJS, withCom = true] failed: java.util.concurrent.TimeoutException: Futures timed out after [1 minute], took 60.143 sec
[error]     at scala.concurrent.impl.Promise$DefaultPromise.ready(Promise.scala:255)
[error]     at scala.concurrent.impl.Promise$DefaultPromise.result(Promise.scala:259)
[error]     at scala.concurrent.Await$.$anonfun$result$1(package.scala:215)
[error]     at scala.concurrent.BlockContext$DefaultBlockContext$.blockOn(BlockContext.scala:53)
[error]     at scala.concurrent.Await$.result(package.scala:142)
[error]     at org.scalajs.jsenv.test.TestKit.awaitAfterClose(TestKit.scala:42)
[error]     at org.scalajs.jsenv.test.TestKit$RunMatcher.hasOutput(TestKit.scala:59)
[error]     at org.scalajs.jsenv.test.RunTests.allowScriptTags(RunTests.scala:42)
[error]     ...
```

sjrd commented:
> I currently get intermittent failures like this on Travis :s

Looks like I tamed that with the latest commit. Now state management is very similar to how it's done in Node.js' com runs.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,315 @@
+        state = AwaitingConnection(sendQueue)
+        val comSetup = makeComSetupFile()
+        val underlyingRun = startRun(comSetup)
+        underlyingRun.future.onComplete { result =>
```
How do you call `close()` on the underlying run?

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,315 @@
+    }
+  }
+
+  private def makeComSetupFile(): VirtualBinaryFile = {
```
Maybe pass `serverPort` as a parameter so you can move this out of the class into the companion?

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,315 @@
+    val listener = new WebsocketListener {
+      def onRunning(): Unit = onServerRunning()
+      def onOpen(): Unit = onConnectionOpened()
+      def onClose(): Unit = ()
```
You must guard completion of the com run by the `WebsocketManager` closing. Otherwise there might be messages delivered after `future` has completed. This is not allowed.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,315 @@
+      |  };
+      |  websocket.onerror = function(evt) {
+      |    websocket = null;
+      |    throw new Error("Websocket failed: " + evt);
```
Does this terminate PhantomJS? If not we should `callPhantom`.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,315 @@
+object ComRun {
+  private final val MaxByteMessageSize = 32768 // 32 KB
+  private final val MaxCharMessageSize = MaxByteMessageSize / 2 // 2B per char
+  private final val MaxCharPayloadSize = MaxCharMessageSize - 1 // frag flag
```
It is fine with me if you retain the whole fragmentation stuff, but if we are OK with upgrading to PhantomJS 2 and Jetty9, this is probably not necessary anymore.

sjrd replied:
We could do that, but it's definitely out of the scope of this PR.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -133,23 +142,28 @@ private final class ComRun(jettyClassLoader: ClassLoader,
-
-      case _ =>
-        throw new AssertionError("Bad fragmentation flag in " + frag)
+    if (!promise.isCompleted) {
```
This has a risk of cutting of trialing messages. For example imagine the code:

```javascript
send("a");
send("b");
send("c");
send("d");
send("e");
exit();
```

If the JVM side drops messages after the run is completed, it might only call `onMessage` for say `a-d`.

sjrd replied:
That risk also exists if we wait for the connection to be closed, AFAICT, because when the script terminates, it can hard-cut the WebSocket connection, and messages could be lost as well.

I first tried to wait for the connection to be closed, but the `onClose()` event was not reliable *at all*, so I would often wait indefinitely. It seems that `onClose()` is in high danger of not being called if `mgr.stop()` was already called.

gzm0 replied:
Ugh :( That is awful. Fair enough. Maybe drop a comment about this. IIUC this also does not cleanly satisfy the no `onMessage` in flight anymore when `future` is completed. But maybe we can live with that for now to unblock everything.

sjrd replied:
I'm still trying something with a new `mgr.close()` that calls `webSocketConn.close()` internally, see if that works better. Otherwise I'll write a comment, yes.

gzm0 replied:
SG

sjrd replied:
OK I think I've found a solution, in the latest commit.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/PhantomJSEnv.scala:
```diff
@@ -215,50 +216,33 @@ final class PhantomJSEnv(config: PhantomJSEnv.Config) extends JSEnv {
-              s"""<script type="text/javascript" src="$fname"></script>""" + "\n")
-      }
+    for (script <- scriptsAndSyntaxErrorHacks) {
+      val scriptURI = materialize(script)
```
I think with these changes, `materialize` always takes `MemVirtualBinaryFile`. Maybe simplify it in light of that?

sjrd replied:
Done

gzm0 commented on phantomjs-env/src/test/scala/org/scalajs/jsenv/phantomjs/PhantomJSSuite.scala:
```diff
@@ -0,0 +1,16 @@
+  val Config = {
+    JSEnvSuiteConfig(new PhantomJSEnv)
+      .withTerminateVMJSCode("__ScalaJSEnv.exitFunction(0)")
+      .withSupportsTimeout(false)
```
Why this? I assume the tests are too restrictive? I also have trouble with the timeout tests on Selenium. So probably we need to improve the test suite there.

sjrd replied:
I'm not sure, I haven't investigated more than the fact that some tests failed, since the status before that PR was also not to execute the Timeout tests:
https://github.com/scala-js/scala-js-env-phantomjs/blob/6a1e55846b84b20dc51de7b9693db78feea1b50c/phantomjs-env/src/test/scala/org/scalajs/jsenv/phantomjs/PhantomJSTest.scala#L7
only extends `ComTests`, not `TimeoutComTests`.

gzm0 replied:
OK, fair enough.

gzm0 replied:
Maybe create an issue and link it here to enable this, once scala-js/scala-js#3420 is fixed?

sjrd replied:
Done (#20).

sjrd commented:
Updated. I think I've addressed all the comments, except the thing about logging. But if you review with `?w=1` in the URL, you'll see that I haven't actually changed that code. It was like that before.

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,371 @@
+        // Wait until the connection is closed before completing the promise
+        connectionClosed.future.foreach { _ =>
+          promise.tryComplete(result)
+        }
```
Couldn't you do this in the `onConnectionClosed` callback instead of having an additional promise?

sjrd replied:
No. There is no guarantee about the order in which `onConnectionClosed()` and `onUnderlyingRunTerminated()` will be executed. So we need this to happen when both have been called, with the `result` received here. The additional promise is the easiest way to do it, although of course I could also use another field with a few states, and the second one that is called triggers the `promise.tryComplete`. Would you prefer that?

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/ComRun.scala:
```diff
@@ -0,0 +1,371 @@
+      case Closing =>
+        // Ignore
+
+      case AwaitingConnection(_) | Connected | NotConnectedAnymore =>
```
`NotConnectedAnymore` seems to never be set. Am I missing something?

sjrd replied:
Indeed. Not sure how that happened. I'll have to revisit when I get the chance. (I'm on vacation until August 13.)

sjrd replied:
Taken care of. See the new commit (to be squashed later).

gzm0 commented on phantomjs-env/src/main/scala/org/scalajs/jsenv/phantomjs/JettyWebsocketManager.scala:
```diff
@@ -132,4 +132,12 @@ private[phantomjs] final class JettyWebsocketManager(
       webSocketConn.sendMessage(msg)
   }
 
+  def closeConnection(): Unit = {
```
I don't fully understand why we need this. Shouldn't `stop()` do that as well?

sjrd replied:
`stop()` is way too abrupt. It kills the connection without giving a chance to the `close` event to be triggered on the JS side. This is necessary for the normal termination to happen on the JS side.

sjrd commented:
> I'm OK with going forward without this, but we should keep this in mind if we get bug reports.

I suggest we go ahead with the status quo. This PR is the last thing blocking the announcement of Scala.js 1.0.0-M5. This PR is about upgrading to M5, not fixing everything that didn't work before (otherwise we'll never get it done).
